unsubscribe-automation/src/main.ts

```diff
@@ -0,0 +1,199 @@
+      Provide a JSON object with:
+      1. An 'actions' array containing steps to unsubscribe. Each action should have:
+         - 'type': Either 'click', 'fill', 'select', or 'submit'
+         - 'selector': A CSS selector for the element. Use only standard CSS selectors (e.g., '#id', '.class', 'tag', '[attribute]').
+         - 'value': (Optional) For input fields. Omit this field if not applicable.
+      2. A 'confirmationIndicator' string to verify success. This should be a CSS selector for an element that appears after successful unsubscription. If uncertain, set to null.
+
+      Return ONLY the JSON object, without any markdown formatting, code blocks, or explanation.
+
+      HTML Content:
+      ${pageContent}
+    `;
+
+  const result = await model.generateContent(prompt);
+  const response = result.response;
+  let analysisText = response.text().trim();
+
+  try {
+    // Remove any markdown code block indicators
+    analysisText = analysisText.replace(/```json\n?|\n?```/g, "");
+
+    // Trim any leading or trailing whitespace
+    analysisText = analysisText.trim();
+
+    const parsedAnalysis = JSON.parse(analysisText);
+    return PageAnalysisSchema.parse(parsedAnalysis);
+  } catch (error) {
+    console.error("Error parsing AI response:", error);
+    console.error("Raw AI response:", analysisText);
+    throw new Error("Failed to parse AI response");
+  }
+}
+
+async function performUnsubscribeActions(
+  page: Page,
+  actions: PageAnalysis["actions"],
+) {
+  for (const action of actions) {
+    console.log(`Attempting action: ${action.type} on ${action.selector}`);
+    try {
+      const locator = page.locator(action.selector);
+
+      if ((await locator.count()) === 0) {
+        console.warn(`Element not found: ${action.selector}`);
+        continue;
+      }
+
+      if (!(await locator.isVisible())) {
+        console.warn(`Element not visible: ${action.selector}`);
+        continue;
+      }
+
+      switch (action.type) {
+        case "click":
+        case "submit":
+          await locator.click({ timeout: 5000 });
+          break;
+        case "fill":
+          if (action.value) {
+            await locator.fill(action.value, { timeout: 5000 });
+          }
+          break;
+        case "select":
+          if (action.value) {
+            await locator.selectOption(action.value, { timeout: 5000 });
+          }
+          break;
+      }
+      console.log(`Action completed: ${action.type} on ${action.selector}`);
+    } catch (error) {
+      console.warn(
+        `Failed to perform action: ${action.type} on ${action.selector}. Error: ${error}`,
+      );
+    }
+    await page
+      .waitForLoadState("networkidle", { timeout: 5000 })
+      .catch(() => {});
+  }
+}
+
+export async function autoUnsubscribe(url: string): Promise<boolean> {
+  // Remove headless: false if you don't want the browser popup to open
+  const browser = await chromium.launch({ headless: false });
```

Comment:
Can set this as an env var. In prod we will likely use headless: true. But it is easier in dev to see it headful mode.

Reply:
Added an env variable